src/tools/wix/BinderFileManager.cs

```diff
@@ -735,7 +777,60 @@ private static bool CheckFileExists(string path)
             }
         }
 
+        /// <summary>
+        /// Private method to retry a file action that may block because
+        /// another process is working the file.  Retries on
+        /// an <see cref="IOException" />.
+        /// </summary>
+        /// <param name="path">File path to watch.</param>
+        /// <typeparam name="T">Return type of the file action.</typeparam>
+        /// <param name="func">File <c>I/O</c> Delegate to retry.</param>
+        /// <returns>
+        /// Returns the result of the delegate on success.
+        /// </returns>
+        private T RetryFileAction<T>(string path, Func<T> func)
+        {
+            var autoResetEvent = new AutoResetEvent(false);
+            int retries = 3;
+
+            while (true)
+            {
+                try
+                {
+                    return func();
+                }
+                catch (IOException)
+                {
+                    var fsw = new FileSystemWatcher(Path.GetDirectoryName(path)) { EnableRaisingEvents = true };
```

Comment:
This can be done by a wait-and-retry method too.  Also, a ceiling on the number of tries can be done easily enough.  The FileSystemWatcher just takes the constants out of the equation.

 

I’m certainly open to other techniques to block while another process is writing—this is just one of many.

 

This code is being tested on four servers with four cores, four agents, and 16 GB memory each supporting five teams with twenty-five products—all migrated to MultiProc.  That works out to between 30-50 builds per day.  There hasn’t been a non-code-non-unit test failure in 5 days.

 

It may not be the best, but it does work in our environment.

 

--

John M. Cooper

Leavenworth, KS

 

 

 

From: Bob Arnson [mailto:notifications@github.com] 
Sent: Sunday, June 28, 2015 3:00 PM
To: wixtoolset/wix3
Cc: John Merryweather Cooper
Subject: Re: [wix3] WIXBUG 4791: (redone) Fix race in lit.exe/light.exe related to wixlibs and la… (#274)

 

In src/tools/wix/BinderFileManager.cs <https://github.com/wixtoolset/wix3/pull/274#discussion_r33426311> :

> +        /// Returns the result of the delegate on success.
> +        /// </returns>
> +        private T RetryFileAction<T>(string path, Func<T> func)
> +        {
> +            var autoResetEvent = new AutoResetEvent(false);
> +            int retries = 3;
> +
> +            while (true)
> +            {
> +                try
> +                {
> +                    return func();
> +                }
> +                catch (IOException)
> +                {
> +                    var fsw = new FileSystemWatcher(Path.GetDirectoryName(path)) { EnableRaisingEvents = true };

This seems pretty heavyweight. Plus it ends up waiting forever instead of timing out, if there's no event on the file. What if it's a file that's causing the problem but we lock on the directory? Recommend omitting.

—
Reply to this email directly or view it on GitHub <https://github.com/wixtoolset/wix3/pull/274/files#r33426311> .  <https://github.com/notifications/beacon/AGu1PV3cpgftOTsL7hmRayyLLmjZ3IVSks5oYEmsgaJpZM4FK_fk.gif> 



---
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus